### ISSUE TITLE
Feature Req: `plt.subplots(sharev=True)`

### ISSUE THREAD

**Issue by jona-sassenhagen** [matplotlib/matplotlib#11377]
### Feature Request

**Summary**

`plt.subplots` has `sharex` and `sharey` parameters. It would be excellent if it also had a `sharev` parameter for aligning color scales across multiple heatmaps.

**Code for reproduction**

```
import numpy as np
import matplotlib.pyplot as plt
_, axes = plt.subolots(nrows=3, sharev=True)
for ax in axes.flat:
    ax.imshow(np.random.randn((10, 10)))
```

**Real-world examples**

For a related issue on stackoverflow, see:
https://stackoverflow.com/questions/17989917/imshow-subplots-with-the-same-colorbar

Example where I've personally come across the issue:
https://mne-tools.github.io/dev/auto_examples/visualization/plot_roi_erpimage_by_rt.html

**Comment by story645** [comment]
I maybe unclear about what you're trying to do. Would creating one [colormap normalization](https://matplotlib.org/users/colormapnorms.html) object that you then pass into all your imshows/matshows/colormaps/etc solve your problem?

**Comment by ImportanceOfBeingErnest** [comment]
The problem with this requested feature is that the `norm` is a property of the artist, not the axes. I.e. it is perfectly possible to create two images with different norms in the same axes. 

So imagine you have a `sharev` argument and set it to True. What should happen if you create images with different norms in the axes? E.g. one image with a linear `Normalize` and one image with a `PowerNorm`? Which one should take precedence?

**Comment by story645** [comment]
>but in the colormap range for the images (not sure if that's the correct term?), i.e., what could be achieved by manually setting vmin and vmax. E.g.,

But that is what the norm controls...the [Normalize](https://matplotlib.org/api/_as_gen/matplotlib.colors.Normalize.html#matplotlib.colors.Normalize) function's signature is 
`matplotlib.colors.Normalize(vmin=None, vmax=None, clip=False)` You set vmin & vmax explicitly in imshow or via a normalization object-and I'd argue that better practice would be to only allow vmin/vmax to be set via a norm object. The colorbar is mostly just a reflection/legend based on what what norm and cmap was set for that axes. 

Basically, I'm advocating for the following:
```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.colors as mcolors
_, axes = plt.subolots(nrows=3, sharev=True)
norm = mcolors.Normalize(vmin=3, vmax=10)
for ax in axes.flat:
    ax.imshow(np.random.randn((10, 10)), norm=norm)
```
`The norms don't seem so problematic to me - could it not simply be made illegal to manually pass norms?`
I think that would be counterproductive to your aims.

**Comment by jona-sassenhagen** [comment]
>(you may have to precompute the min/max across all the images you're plotting):

That is exactly what I'm trying to avoid :( I'm looking at a situation where pre-computing the vlims can be very complicated. I.e., I have a bit of code where this means I have to do the same loop twice, once to determine the max lims, a second to do the actual plotting.

I don't see how the norm object helps in that case. Once I have computed the max (or otherwise optimal) lims across all plots, everything is easily solved by forwarding the max lims to `imshow`.

Unless I'm still misunderstanding you ...

So your code example doesn't solve my problem; I'm looking for something that dynamically adapts lims based on all plotted images (rather than, e.g., some global feature of the underlying data).

Sorry if that's still unclear.

**Comment by afvincent** [comment]
@jona-sassenhagen By any chance, would the following [example](https://matplotlib.org/gallery/images_contours_and_fields/multi_image.html) (or at least its first half) be close to what you want to do?

**Comment by jona-sassenhagen** [comment]
@afvincent 

That is what I'm doing right now :) I'd wish there was some functionality to automate that part, for much cleaner and simpler user-facing code.

**Comment by timhoffm** [comment]
What you already can do is this:

~~~
import matplotlib.pyplot as plt
from matplotlib.colors import Normalize

fields = [np.random.randn(10, 10) + 2*i for i in range(3)]

_, axes = plt.subplots(ncols=3)
norm = Normalize()
norm.autoscale(fields)
for ax, field in zip(axes, fields):
    ax.imshow(field, norm=norm, cmap="RdBu_r")
plt.colorbar(im, ax=axes)
~~~

![grafik](https://user-images.githubusercontent.com/2836374/41001137-61aff5fe-6910-11e8-9340-cc0a6919d61b.png)

We could slightly simplify this by adding a parameter  `Normalize(autoscale=fields)`.

Alternatively, one could define a class method `Normalize.apply_autoscale(images)` that takes a list of images, creates an autoscaled `Normalize` instance and applies it to all the images. That would look like
~~~
images = []
for ax, field in zip(axes, fields):
    images.append(ax.imshow(field, cmap="RdBu_r"))
Normalize.apply_autoscale(images)
~~~

**Comment by tacaswell** [comment]
> That is exactly what I'm trying to avoid :( I'm looking at a situation where pre-computing the vlims can be very complicated. I.e., I have a bit of code where this means I have to do the same loop twice, once to determine the max lims, a second to do the actual plotting.

You can acumulate the min/max and do the plotting on one loop and then do the update at the end (either with a loop over the artists or my updating a shared `Normalize` object).

You can stuff all of that logic into a function, ex

```python
def plot_all_my_images_shared_limits(list_of_axes, list_of_images, **kwargs):
    imgs = []
    vmin, vmax = np.inf, -np.inf
    for ax, data in zip(list_of_axes, list_of_images):
        # or what ever computation you need to do
        vmin = min(vmin, np.min(data))
        vmax = max(vmax, np.max(data))
        imgs.append(ax.imshow(data, **kwargs)

    for im in imgs:
         im.set_clim(vmin, vmax)
    return imgs
```

and now you don't even have to see the loop in your code :wink: 
 
Another option is 

```python
def imshow_update_norm(ax, data, norm, **kwargs):
    norm.vmin = min(np.min(data), norm.vmin) if norm.vmin is not None else np.min(data)
    norm.vmax = max(np.max(data), norm.vmax) if norm.vmax is not None else np.max(data)
    return ax.imshow(data, norm=norm, **kwargs)
```

> I'm looking for something that dynamically adapts lims based on all plotted images (rather than, e.g., some global feature of the underlying data).

This request requires putting a lot of image-specific (or at least colormappable specfiic) state into the axes objects.  While this would make this specific use case terser, I do not think that is worth the complexity we would have to pick up (particularly given that you can write a (terser to use) function even more tailored to your needs).  

If we supported the color limits like this, would we also want to support interpolation, extent, ... this way too?  

What should happen if someone want to put two colormappable objects with different limits into the same Axes?

I am going to close this as wont-fix and discussion should move over to #11380 .

**Comment by jona-sassenhagen** [comment]
I understand. Thanks!
